Add verified contract

greenhat opened the pull request:
Todo:
- [x] in `AssetsAtomicExchangeSellerCmd` use `BoxOperations.selectTop` that searches for tokens;
- [x] in `AssetsAtomicExchangeMatchCmd` check that passed `sellerAddress` is in the seller's contract ErgoTree; 
- [x] in `AssetsAtomicExchangeMatchCmd` check that passed `buyerAddress` is in the buyer's contract ErgoTree; 
- [x] in `AssetsAtomicExchangeMatchCmd` check that token (`box.tokens`) from seller's contract box is in the buyer's contract ErgoTree; 
- [x] in `AssetsAtomicExchangeMatchCmd` check that coins (`box.value`) from buyer's contract box are in seller's contract ErgoTree; 
- [ ] implement Cmd that lists Bids and Asks with Options to list only valid matches;
- [ ] test list cmd;
- [x] move commands under AssetsAtomicExchange package;
- [ ] test on testnet;
- [ ] test on mainnet;

greenhat committed:
draft AssetsAtomicExchangeSellerCmd;
draft a test in ErgoToolSpec for AssetsAtomicExchangeSellerCmd;
add token to seller's contract;

greenhat committed:
draft AssetsAtomicExchangeBuyerCmd;
formatting;

aslesarenko commented on src/main/scala-2.12/org/ergoplatform/appkit/ergotool/AssetsAtomicExchangeBuyerCmd.scala:
```diff
@@ -0,0 +1,95 @@
+    val status = if (res != null) "Ok" else "Error"
+    console.println(s" $status")
+    res
+  }
```
this is now used in more than one command, let's refactor.

greenhat replied:
Moved to `Cmd` in https://github.com/aslesarenko/ergo-tool/pull/3/commits/e93ca8abe16eb9261b29c2c70b7d62abbc693541

greenhat committed:
draft AssetsAtomicExchangeMatchCmd;
comment;

greenhat committed:
add scaladoc to AssetsAtomicExchange*Cmd;

greenhat committed:
add coin selection for tx fee in AssetsAtomicExchangeMatchCmd;

greenhat committed:
removed token parameter from AssetsAtomicExchangeMatchCmd (get from seller's contract box);

greenhat committed:
update appkit and use ErgoId.create;

greenhat committed:
add spec test for AssetsAtomicExchangeMatchCmd;
add todo;

greenhat committed:
use selectTop with token in AssetsAtomicExchangeSellerCmd;

greenhat committed:
check seller's address and token price are in seller's contract constants;

greenhat commented:
Done:
- check the token amount in `BoxOperations.selectTop` and use it in `AssetsAtomicExchangeSellerCmd`;
- added checks in `AssetsAtomicExchangeMatchCmd` that passed `sellerAddress` and ergo price are in the seller's contract ErgoTree constants;

aslesarenko commented:
@greenhat I think we need to generalize original contracts:
- introduce DEX role. 
- DEX is given by pkDEX. 
- Each original order transaction besides creating the box for matching, also creates the box for pkDEX to spend, with additional condition that this spending happens in the same transaction together with spending of buyer and seller boxes.
- We can call these two new boxes and the contract DexFee. DexFee is payed by both seller and buyer. The goal is to incentivize DEX service.

greenhat commented:
@aslesarenko I agree. When I created matching tx it struck me that DEX has to pay a tx fee, which is discouraging. We should put it into an issue (in sigmastate repo?).

greenhat committed:
check buyer's address is in buyer's contract constants;

greenhat committed:
check token id and amount are in buyer's contract constants;
update appkit (new sigma);

greenhat commented:
@aslesarenko I've tried to implement DexFee and it came up larger than I expected:
```
  def buyerDexFee(ctx: Context,
                  pkDex: SigmaProp,
                  pkBuyer: SigmaProp,
                  tokenId: Coll[Byte],
                  tokenAmount: Long): SigmaProp = {
    import ctx._
    pkDex &&
      (OUTPUTS.nonEmpty && OUTPUTS(0).R5[Coll[Byte]].isDefined) && {
      val tokens = OUTPUTS(0).tokens
      val tokenDataCorrect = tokens.nonEmpty &&
        tokens(0)._1 == tokenId &&
        tokens(0)._2 >= tokenAmount

      val knownId = OUTPUTS(0).R5[Coll[Byte]].get == SELF.id
      tokenDataCorrect &&
        OUTPUTS(0).propositionBytes == pkBuyer.propBytes &&
        knownId
    }
  }
}
```
I could not make a shorter code for a "that this spending happens in the same transaction together with spending of buyer and seller boxes" condition. I find it way too complicated.

I believe I found a much simpler solution to DEX's fee. Basically buyer's fee to DEX can be `buyerHolderBox.value - sellerContract.ergAmount` and seller's fee to DEX can be the whole `sellerHolderBox.value`. This way, no changes to contracts are needed, only a change to the DEX matching algorithm. Extraction of the `sellerContract.ergAmount` from seller's contract ErgoTree seems doable.

aslesarenko commented:
@greenhat It shouldn't depend on matching algorithm. Basically DexFee can be decided for each order transaction. The patterns should be similar to the ErgoScriptPredef.feeProposition

greenhat commented:
@aslesarenko Could you please clarify. Why it should not depend on matching algorithm? 
I find the DEX fee `buyerHolderBox.value - sellerContract.ergAmount + sellerHolderBox.value` quite incentivizing.

aslesarenko commented:
@greenhat what is sellerContract.ergAmount, where is come from?

greenhat commented:
@aslesarenko 

> @greenhat what is sellerContract.ergAmount, where is come from?

Amount of coins seller wants for their tokens. Parsed by DEX matching algorithm from the seller contract holder box (passed as `sellerHolderBoxId` to DEX).

greenhat commented:
Not fixing the on DEX (with pk) in buyer's and seller's contract gives a chance to anyone to match them, thus a free market will determine DEX fees.

greenhat committed:
add DEX fee as (buyerHolderBox.value - sellerContract.askedErgAmount + sellerHolderBox.value - MinFee);

greenhat committed:
add a check for not enough value in buyers contract to close the seller's contract in DEX matching;
naming;

aslesarenko commented on src/main/scala-2.12/org/ergoplatform/appkit/ergotool/AssetsAtomicExchangeMatchCmd.scala:
```diff
@@ -76,22 +75,28 @@ case class AssetsAtomicExchangeMatchCmd(toolConf: ErgoToolConfig,
-      val boxesForTxFee = BoxOperations.selectTop(unspent, MinFee)
-      boxesForTxFee.addAll(util.Arrays.asList(buyerHolderBox, sellerHolderBox))
-      val inputBoxes = boxesForTxFee
+      val ergAmount = sellerHolderBox.getErgoTree.constants(6).asInstanceOf[Constant[SLong.type]].value
```
Well, in this case other parameters can also be extracted from the contract.

greenhat replied:
Good point! Removed buyer and seller addresses from the parameters in https://github.com/aslesarenko/ergo-tool/pull/3/commits/d197c510bab8c77896f28640d15ce3e104275eb9

aslesarenko commented on src/main/scala-2.12/org/ergoplatform/appkit/ergotool/AssetsAtomicExchangeMatchCmd.scala:
```diff
@@ -76,22 +75,28 @@ case class AssetsAtomicExchangeMatchCmd(toolConf: ErgoToolConfig,
-      val inputBoxes = boxesForTxFee
+      val ergAmount = sellerHolderBox.getErgoTree.constants(6).asInstanceOf[Constant[SLong.type]].value
+
+      val claimableValue = buyerHolderBox.getValue - ergAmount + sellerHolderBox.getValue - 1 // 1 for buyerTokensOutBox.value
```
What  claimable means? Also this formula is not readable. What -1 means?

greenhat replied:
`claimableValue` is (total input - total outputs). Deduct `MinFee` from it and it will be DEX fee. We need a better name. :)

aslesarenko commented on src/main/scala-2.12/org/ergoplatform/appkit/ergotool/AssetsAtomicExchangeMatchCmd.scala:
```diff
@@ -76,22 +75,28 @@ case class AssetsAtomicExchangeMatchCmd(toolConf: ErgoToolConfig,
       val txB = ctx.newTxBuilder
       val buyerTokensOutBox = txB.outBoxBuilder
-        .value(sellerHolderBox.getValue)
+        .value(1)
```
MinFee is minimal value that can be put in the box. 1 NanoErg is too small.
See org.ergoplatform.appkit.Parameters

greenhat replied:
Thanks. I missed that here. Fixed in https://github.com/aslesarenko/ergo-tool/pull/3/commits/424752c24d123e7cfe7d18124667a64f7d087d76

greenhat committed:
move loggedStep to Cmd;

greenhat committed:
fix minimal value in buyer out box in AssetsAtomicExchangeMatchCmd;

greenhat committed:
extract seller's and buyer's addresses for matching transaction from their contracts;

greenhat committed:
extract txFee, better naming;

greenhat committed:
drafting AssetsAtomicExchangeListCmd;
update appkit;
use ErgoTreeTemplate;

greenhat committed:
draft list cmd;
extract contract specific code into AssetsAtomicExchange*Contract;
add console output for the found matching contracts;

greenhat commented:
- drafted contract matching - [src](http://github.com/aslesarenko/ergo-tool/blob/1ccf7ba5525c3a59bc8952925f73d22d742b4f71/src/main/scala-2.12/org/ergoplatform/appkit/ergotool/AssetsAtomicExchangeListCmd.scala#L76-L93)
- extracted contract specific code into `AssetsAtomicExchange*Contract` objects [buyer](http://github.com/aslesarenko/ergo-tool/blob/1ccf7ba5525c3a59bc8952925f73d22d742b4f71/src/main/scala-2.12/org/ergoplatform/appkit/ergotool/AssetsAtomicExchangeBuyerCmd.scala#L114-L133), [seller](http://github.com/aslesarenko/ergo-tool/blob/1ccf7ba5525c3a59bc8952925f73d22d742b4f71/src/main/scala-2.12/org/ergoplatform/appkit/ergotool/AssetsAtomicExchangeSellerCmd.scala#L113-L130);

greenhat committed:
move AssetsAtomicExchange commands into the separate package;
unused imports;
update scaladoc;

greenhat committed:
draft tests for listing matching contracts;

greenhat committed:
update CreateBuyerContractCmd scaladoc;
comments;

greenhat commented:
- adding tests for matching contracts cmd;

@aslesarenko I've added `MockInputBox` and started writing generators for property-based tests - [src](http://github.com/aslesarenko/ergo-tool/blob/515b913257943ade07ed147390095a021320e028/src/test/scala/org/ergoplatform/appkit/ergotool/AssetsAtomicExchange/ListMatchingContractsSpec.scala#L19-L39);

greenhat commented:
List cmd continue in #10
